pkg/relay/relay.go

```diff
@@ -0,0 +1,513 @@
+	}
+	timer2.ObserveDuration()
+
+	header := headers[0] // choose the highest bid, which is index 0
+
+	if header.Header == nil || (header.Header.ParentHash != parentHash) {
+		log.Debug(badHeaderMsg)
+		return nil, fmt.Errorf(noBuilderBidMsg)
+	}
+
+	bid := types.BuilderBid{
+		Header: header.Header,
+		Value:  header.Trace.Value,
+		Pubkey: rs.config.PubKey,
+	}
+
+	signature, err := types.SignMessage(&bid, rs.config.BuilderSigningDomain, rs.config.SecretKey)
+	if err != nil {
+		return nil, fmt.Errorf("internal server error")
+	}
+
+	/*
+		logger.With(log.F{
+			//"processingTimeMs": time.Since(timeStart).Milliseconds(),
+			"bidValue":     bid.Value.String(),
+			"blockHash":    bid.Header.BlockHash.String(),
+			"feeRecipient": bid.Header.FeeRecipient.String(),
+			"slot":         slot,
+		}).Trace("bid sent")*/
+
+	return &types.GetHeaderResponse{
+		Version: "bellatrix",
+		Data:    &types.SignedBuilderBid{Message: &bid, Signature: signature},
+	}, nil
+}
+
+// GetPayload is called by a block proposer communicating through mev-boost and reveals execution payload of given signed beacon block if stored
+func (rs *Relay) GetPayload(ctx context.Context, payloadRequest *types.SignedBlindedBeaconBlock) (*types.GetPayloadResponse, error) { // TODO(l): remove FB type
+	timer := prometheus.NewTimer(rs.m.Timing.WithLabelValues("getPayload", "all"))
+	defer timer.ObserveDuration()
+
+	logger := rs.l.WithField("method", "GetPayload")
+
+	if len(payloadRequest.Signature) != 96 {
+		return nil, fmt.Errorf("invalid signature")
+	}
+
+	proposerPubkey, err := rs.beaconState.Beacon().KnownValidatorByIndex(payloadRequest.Message.ProposerIndex)
+	if err != nil && errors.Is(err, structs.ErrUnknownValue) {
+		return nil, fmt.Errorf("unknown validator for index %d", payloadRequest.Message.ProposerIndex)
+	} else if err != nil {
+		return nil, err
+	}
+
+	timer2 := prometheus.NewTimer(rs.m.Timing.WithLabelValues("getPayload", "verify"))
+	pk, err := types.HexToPubkey(proposerPubkey.String())
+	if err != nil {
+		return nil, err
+	}
+	logger.With(log.F{
+		"slot":      payloadRequest.Message.Slot,
+		"blockHash": payloadRequest.Message.Body.ExecutionPayloadHeader.BlockHash,
+		"pubkey":    pk,
+	}).Debug("payload requested")
+
+	msg, err := types.ComputeSigningRoot(payloadRequest.Message, rs.config.ProposerSigningDomain)
+	if err != nil {
+		return nil, fmt.Errorf("signature invalid") // err
+	}
+
+	respCh := rs.singleRetChannPool.Get().(chan SVRReqResp)
+	rs.regMngr.VerifyChanStacks(ResponseQueueOther) <- SVRReq{
+		Signature: payloadRequest.Signature,
+		Pubkey:    pk,
+		Msg:       msg,
+		Response:  respCh}
+	resp := <-respCh
+	rs.singleRetChannPool.Put(respCh)
+	timer2.ObserveDuration()
+	if resp.Err != nil {
+		logger.WithField(
+			"pubkey", proposerPubkey,
+		).Error("signature invalid")
+		return nil, fmt.Errorf("signature invalid")
+	}
+
+	timer3 := prometheus.NewTimer(rs.m.Timing.WithLabelValues("getPayload", "getPayload"))
```

Comment:
Are we planning to leave the timers? Or is it just for performance analysis?

Reply:
I was suspicious, the times taking so long,  so I put them there. It wouldn't hurt to left it here, but totally fine to remove /rename them as well

Reply:
Okay, so it really depends on how much value these timers you consider they add. If you think they may be useful for future peformance improvements we leave them. Otherwise we just leave the outer timer. Your choice!